BUG: Mismatched allocation domains in ``PyArray_FillWithScalar``

rostan-t opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->
`PyArray_FillWithScalar` allocates memory with `PyObject_Calloc` but releases it with `PyMem_FREE`.

rostan-t committed:
TST: add a failing test for ndarray.fill

When `PyArray_FillWithScalar` needs to heap-allocate a buffer, the
deallocating function does not use the same domain as the allocating
function. Such API violations are detected when debug hooks are enabled.

rostan-t committed:
BUG: use the "Mem" domain for allocation in `PyArray_FillWithScalar`

seberg commented:
Thanks for finding and addressing this!  I guess we probably really not have a test that triggers this.
Can you modify the test to just add a comment that it is also a regression test for a crash with `PYTHONMALLOC=debug`?  Changing the environment variable doesn't actually work, as python reads it only at startup.  It would be nice to make sure one CI job uses it, but I happy to not include that here.

rostan-t committed:
TST: do not explicitely set PYTHONMALLOC in test_fill_readonly

The test now only mentions that it tests a regression when the
variable is set to debug.

rostan-t commented:
Thanks for the feedback! de0046f removes `PYTHONMALLOC` from the test.

seberg commented:
Thanks @rostan-t.
